### DISCUSSION TITLE
Add Objects (PDFs, Excel, Image, Word etc) and update the icon image and captions programmatically using OpenXML

### DISCUSSION THREAD

**Discussion by AlgorithmicThoughts** [OfficeDev/Open-XML-SDK#1830] [General]
I'm working on an ASP.NET project where I am developing an endpoint that uses OpenXML to generate an Excel file. Additionally, I create a PDF document using a third-party tool. I need to insert the PDF stream into an Excel sheet (similar to insert object option of the excel). I discovered [via a StackOverflow post](https://stackoverflow.com/questions/3322247/how-can-i-embed-any-file-type-into-microsoft-word-using-openxml-2-0) that this can be done using COM Interop which I want to avoid as this would mean storing the Excel DLLS on the server and starting a new instance of excel whenever I want to add the PDFs into the excel (please correct if my understanding is incorrect). 

I would also like to update the icon and the caption of the attached PDF object (this is also a requirement I am currently looking into but was not able to get an answer to)

![image](https://github.com/user-attachments/assets/6b3b5bf2-93ed-4df4-9252-42c29da3aea2)

Is there a way to insert the PDF stream directly into the Excel sheet and update it's icon and caption without relying on COM? Are there existing options to achieve this?

**Describe the solution you'd like**

A method to embed PDF streams and update its icon image and captions into Excel sheets directly using OpenXML, eliminating the need for COM Interop or any third party.

**Here's what I know so far but it does not work in my case**

1. Using COM Interop to embed the PDF, which isn't feasible in my case.
2. Converting the PDF to an image and embedding it, which isn't ideal for maintaining document integrity.
3. Including a hyperlink to the PDF location, which doesn't truly embed the document in the Excel file.

This feature would greatly enhance the flexibility and functionality of OpenXML, providing a more seamless experience for developers needing to embed various objects in Office documents.

**Comment by ThomasBarnekow** [comment]
@AlgorithmicThoughts, do you know exactly what Open XML markup you'll have to generate for that purpose and what other binary parts, if any, will have to be changed to create the desired effect? If you knew what the result should look like, you could build that using the Open XML SDK.

**Reply by AlgorithmicThoughts** [reply]
I tried that with OpenXml @ThomasBarnekow  and took a different route when I realized to achieve this I need an instance of excel running on the server, so we went a different route on this.

**Comment by twsouthwick** [comment]
@AlgorithmicThoughts this is not so much an issue with the SDK but a how-to. Stackoverflow or the discussions here are a better place for that.

**Reply by AlgorithmicThoughts** [reply]
Thanks for the clarification @twsouthwick with the situation I am in I decided not to attach the PDF into the excel because to do that we need to run an instance of the excel on our server which I think is not advisable. So I end up taking a different route of creating zip files with excel and the associated PDFs.

I wish there was better documentation of what OpenXml can or cannot do in simple terms or easy to follow examples or even simple wrappers for adding images than having to write lengthy code. If it wasn't for OpenXml productivity tools I would not have known the amount of code I have to write just to add an image in a cell.

OpenXml is great but the lack of good documentation hurts.

**Reply by ThomasBarnekow** [reply]
@twsouthwick, I think it would really be helpful to have a description of the following:
- What does the Open XML SDK provide at its core: an SDK with APIs for **low-level** operations related to OPC packages, Flat OPC files, and Open XML markup in two forms (i.e., strongly-typed classes and LINQ to XML).
- What does the Open XML SDK **not** provide: higher-level abstractions and productivity tools.
- What are the prerequisites for using the Open XML SDK: detailed knowledge of the relevant formats and standards

Too many developers are struggling because they don't understand Open XML etc. and somehow hope the Open XML SDK makes it easy to work with Office documents (e.g., Word, Excel). The problem is it doesn't, at least not in the way and to the extent developers think it does or should do from their point of view.
